README
API Reference:

- normalize_url(url, charset='utf8'):
    Sometimes you get an URL by a user that just isn't a real
    URL because it contains unsafe characters like ' ' and so on.  This
    function can fix some of the problems in a similar way browsers
    handle data entered by the user:

    >>> normalize_url(u'http://de.wikipedia.org/wiki/Elf (Begriffsklärung)')
    'http://de.wikipedia.org/wiki/Elf%20%28Begriffskl%C3%A4rung%29'

   :param charset: The target charset for the URL if the url was
                   given as unicode string.

- optimize
    Decorator to run a callable through psyco:
   
        >>> @optimize
        ... def sum(a, b):
        ...     return a + b
        ...
        >>> sum(1, 2)
        3

- sh(cmd)
    run cmd in a subprocess and return its output.
    Same effect as running the command in the shell.
    raises RuntimeError on error.

- hilite(string, ok=True, bold=False):
    Return an highlighted version of 'string'.

- profile(sort=('time', 'calls'), lines=30, strip_dirs=False)
    A decorator to profile a callable.

- run_in_thread(*args, **kwargs)
    A decorator to run a function in a thread
    
- ScriptBase
    A base class for import scripts.
    
- compat module:
    import it as such:
    >>> from gerbrandyutils.compat import *
    ...and you'll have all(), any() and collections.namedtuple on python
    versions < 2.6.
    
    


HISTORY
0.1.7
add ScriptBase

0.1.6
add run_in_thread()

0.1.5
add compat module within all(), any() and namedtuple() functions.

0.1.4
add hilite()

0.1.3
add optmize() decorator

0.1.2
add sh()

0.1.1
add normalize_url()

0.1.1
add profile()



gerbrandyutils/tests/test_case.py
#!/usr/bin/env python

import unittest
import time
import threading

from gerbrandyutils import sh, normalize_url, hilite, run_in_thread
from gerbrandyutils.compat import all, any, namedtuple
try:
    from gerbrandyutils import optimize
except ImportError:
    optimize = None


class TestCase(unittest.TestCase):

    def test_sh(self):
        stdout = sh("ls")
        self.assertTrue(stdout)
        self.assertRaises(RuntimeError, sh, 'badcmd')
          
    def test_normalize_url(self):
        nu = normalize_url
        self.assertEqual(nu('http://google.it'), 'http://google.it')
        self.assertEqual(nu('http://google.it?a=1&b=2'), 'http://google.it?a=1&b=2')

    if optimize is not None:
        def test_optimize(self):

            def foo():
                L = []
                for i in range(100):
                    L.append(i)
            t1 = time.time()
            for x in range(10000):
                foo()
            normal_time = time.time() - t1

            @optimize
            def foo():
                L = []
                for i in range(100):
                    L.append(i)
            t1 = time.time()
            for x in range(10000):
                foo()
            optimized_time = time.time() - t1
            self.assertTrue(optimized_time < normal_time)
            
    def test_hilite(self):
        hilite("foo", ok=1)
        hilite("foo", ok=1, bold=1)
        hilite("foo", ok=0)
        hilite("foo", ok=0, bold=1)
        
    def test_compat_all(self):
        self.assertTrue(all([1, 1, 1, 1]))
        self.assertFalse(all([1, 1, 1, 0]))

    def test_compat_any(self):
        self.assertTrue(any([1, 1, 1, 1]))
        self.assertTrue(any([1, 1, 1, 0]))
        self.assertFalse(all([0, 0, 0, 0]))
        
    def test_compat_namedtuple(self):
        init = namedtuple('meminfo', 'rss vms')
        nt = init(1, 2)
        self.assertEqual(nt[0], nt.rss)
        self.assertEqual(nt[1], nt.vms)
        self.assertEqual(str(nt), "meminfo(rss=1, vms=2)")
        
    def test_run_in_thread(self):
        flag = []
        
        @run_in_thread
        def foo():
            time.sleep(0.1)
            flag.append(None)
           
        t = foo()
        self.assertTrue(isinstance, threading.Thread)
        self.assertTrue(t.isAlive)
        t.join()
        self.assertFalse(t.isAlive())
        self.assertEqual(flag, [None])


def test_suite():
    test_suite = unittest.TestSuite()
    tests = [TestCase]
    for test in tests:
        test_suite.addTest(unittest.makeSuite(test))
    return test_suite

if __name__ == "__main__":
    unittest.main(defaultTest='test_suite')    



gerbrandyutils/compat.py
#!/usr/bin/env python

__all__ = ["all", "any", "namedtuple"]

"""
Module which provides compatibility with older Python versions
starting from Python 2.4.
Recommended usage:

from gerbrandyutils.compat import *

...at the top of the module.
"""



from operator import itemgetter as _itemgetter
from keyword import iskeyword as _iskeyword
import sys as _sys


# 2.5
try:
    all = all
except NameError:
    def all(iterable):
        for element in iterable:
            if not element:
                return False
        return True

# 2.5
try:
    any = any
except NameError:
    def any(iterable):
        for element in iterable:
            if element:
                return True
        return False

# 2.6 - 2.7 (rename arg)
try:
    from collections import namedtuple
except ImportError:
    def namedtuple(typename, field_names, verbose=False, rename=False):
        """A collections.namedtuple implementation written in Python.
        Taken from: http://code.activestate.com/recipes/500261/
        """
        # Parse and validate the field names.  Validation serves two
        # purposes, generating informative error messages and preventing
        # template injection attacks.
        if isinstance(field_names, basestring):
             # names separated by whitespace and/or commas
            field_names = field_names.replace(',', ' ').split()
        field_names = tuple(map(str, field_names))
        if rename:
            names = list(field_names)
            seen = set()
            for i, name in enumerate(names):
                if (not min(c.isalnum() or c=='_' for c in name) or _iskeyword(name)
                    or not name or name[0].isdigit() or name.startswith('_')
                    or name in seen):
                        names[i] = '_%d' % i
                seen.add(name)
            field_names = tuple(names)
        for name in (typename,) + field_names:
            if not min(c.isalnum() or c=='_' for c in name):
                raise ValueError('Type names and field names can only contain ' \
                                 'alphanumeric characters and underscores: %r'
                                 % name)
            if _iskeyword(name):
                raise ValueError('Type names and field names cannot be a keyword: %r' \
                                 % name)
            if name[0].isdigit():
                raise ValueError('Type names and field names cannot start with a ' \
                                 'number: %r' % name)
        seen_names = set()
        for name in field_names:
            if name.startswith('_') and not rename:
                raise ValueError('Field names cannot start with an underscore: %r'
                                 % name)
            if name in seen_names:
                raise ValueError('Encountered duplicate field name: %r' % name)
            seen_names.add(name)

        # Create and fill-in the class template
        numfields = len(field_names)
        # tuple repr without parens or quotes
        argtxt = repr(field_names).replace("'", "")[1:-1]
        reprtxt = ', '.join('%s=%%r' % name for name in field_names)
        template = '''class %(typename)s(tuple):
        '%(typename)s(%(argtxt)s)' \n
        __slots__ = () \n
        _fields = %(field_names)r \n
        def __new__(_cls, %(argtxt)s):
            return _tuple.__new__(_cls, (%(argtxt)s)) \n
        @classmethod
        def _make(cls, iterable, new=tuple.__new__, len=len):
            'Make a new %(typename)s object from a sequence or iterable'
            result = new(cls, iterable)
            if len(result) != %(numfields)d:
                raise TypeError('Expected %(numfields)d arguments, got %%d' %% len(result))
            return result \n
        def __repr__(self):
            return '%(typename)s(%(reprtxt)s)' %% self \n
        def _asdict(self):
            'Return a new dict which maps field names to their values'
            return dict(zip(self._fields, self)) \n
        def _replace(_self, **kwds):
            'Return a new %(typename)s object replacing specified fields with new values'
            result = _self._make(map(kwds.pop, %(field_names)r, _self))
            if kwds:
                raise ValueError('Got unexpected field names: %%r' %% kwds.keys())
            return result \n
        def __getnewargs__(self):
            return tuple(self) \n\n''' % locals()
        for i, name in enumerate(field_names):
            template += '        %s = _property(_itemgetter(%d))\n' % (name, i)
        if verbose:
            print template

        # Execute the template string in a temporary namespace
        namespace = dict(_itemgetter=_itemgetter, __name__='namedtuple_%s' % typename,
                         _property=property, _tuple=tuple)
        try:
            exec template in namespace
        except SyntaxError, e:
            raise SyntaxError(e.message + ':\n' + template)
        result = namespace[typename]

        # For pickling to work, the __module__ variable needs to be set
        # to the frame where the named tuple is created.  Bypass this
        # step in enviroments where sys._getframe is not defined (Jython
        # for example) or sys._getframe is not defined for arguments
        # greater than 0 (IronPython).
        try:
            result.__module__ = _sys._getframe(1).f_globals.get('__name__', '__main__')
        except (AttributeError, ValueError):
            pass

        return result



gerbrandyutils/__init__.py
#!/usr/bin/env python
# coding=utf8

__all__ = [# --- modules
           "compat",
           # --- functions
            "normalize_url", "sh", "hilite", 
           # --- decorators
           "profile", "optimize", "run_in_thread",
           # --- classes           
           "ScriptBase"]

import tempfile
import urllib
import urlparse
import os
import sys
import subprocess
import warnings
import threading
import tarfile
import shutil
import time

try:
    import cProfile
    import pstats
except ImportError:
    cProfile = None  # fix it with "sudo apt-get install python-profiler"

try:
    import psyco
except ImportError:
    psyco = None

import compat  # pushes compat namespace up at this level

if cProfile is not None:

    def profile(sort='cumulative', lines=50, strip_dirs=False):
        """A decorator which profiles a callable.
        Example usage:

        >>> @profile
        ... def factorial(n):
        ...     n = abs(int(n))
        ...     if n < 1: 
        ...             n = 1
        ...     x = 1
        ...     for i in range(1, n + 1):
        ...             x = i * x
        ...     return x
        ... 
        >>> factorial(5)
        Thu Jul 15 20:58:21 2010    /tmp/tmpIDejr5

                 4 function calls in 0.000 CPU seconds

           Ordered by: internal time, call count

           ncalls  tottime  percall  cumtime  percall filename:lineno(function)
                1    0.000    0.000    0.000    0.000 profiler.py:120(factorial)
                1    0.000    0.000    0.000    0.000 {range}
                1    0.000    0.000    0.000    0.000 {abs}
                1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

        120
        >>>
        """
        def outer(fun):
            def inner(*args, **kwargs):
                file = tempfile.NamedTemporaryFile()
                prof = cProfile.Profile()
                try:
                    ret = prof.runcall(fun, *args, **kwargs)
                except:
                    file.close()
                    raise

                prof.dump_stats(file.name)
                stats = pstats.Stats(file.name)
                if strip_dirs:
                    stats.strip_dirs()
                if isinstance(sort, (tuple, list)):
                    stats.sort_stats(*sort)
                else:
                    stats.sort_stats(sort)
                stats.print_stats(lines)

                file.close()
                return ret
            return inner

        # in case this is defined as "@profile" instead of "@profile()"
        if hasattr(sort, '__call__'):
            fun = sort
            sort = 'cumulative'
            outer = outer(fun)
        return outer


# TODO - can be extended to support python 2.6 class decorators
if psyco is not None:
    def optimize(fun):
        """Decorator to optimize a callable by using psyco.

        >>> @optimize
        ... def sum(a, b):
        ...     return a + b
        ...
        >>> sum(1, 2)
        3
        >>>
        """   
        def outer(fun):
            def inner(*args, **kwargs):
                psyco.bind(fun)
                return fun(*args, **kwargs)
            return inner
        return outer(fun)


def normalize_url(s, charset='utf-8'):
    """Sometimes you get an URL by a user that just isn't a real
    URL because it contains unsafe characters like ' ' and so on.  This
    function can fix some of the problems in a similar way browsers
    handle data entered by the user:

    >>> normalize_url(u'http://de.wikipedia.org/wiki/Elf (Begriffsklärung)')
    'http://de.wikipedia.org/wiki/Elf%20%28Begriffskl%C3%A4rung%29'

    :param charset: The target charset for the URL if the url was
                    given as unicode string.
    """
    if isinstance(s, unicode):
        s = s.encode(charset, 'ignore')
    scheme, netloc, path, qs, anchor = urlparse.urlsplit(s)
    path = urllib.quote(path, '/%')
    qs = urllib.quote_plus(qs, ':&=')
    return urlparse.urlunsplit((scheme, netloc, path, qs, anchor))


def sh(cmdline):
    """run cmd in a subprocess and return its output.
    raises RuntimeError on error.
    """
    p = subprocess.Popen(cmdline, shell=1, stdout=subprocess.PIPE,
                                           stderr=subprocess.PIPE)
    stdout, stderr = p.communicate()
    if p.returncode != 0:
        raise RuntimeError(stderr)
    if stderr:
        warnings.warn(stderr, RuntimeWarning)
    return stdout


def hilite(string, ok=True, bold=False):
    """Return an highlighted version of 'string'."""
    attr = []
    if ok:  # green
        attr.append('32')
    elif ok != -1:  # red
        attr.append('31')
    if bold:
        attr.append('1')
    return '\x1b[%sm%s\x1b[0m' % (';'.join(attr), string)
    
if not sys.stdout.isatty():
    hilite = lambda string, ok=True, bold=False: string


def run_in_thread(fn):
    """Decorator to run a callable in a thread returning the 
    thread instance.
    Note: completely *NOT* thread safe.

    >>> import time
    >>>
    >>> @run_in_thread
    ... def foo():
    ...     time.sleep(100)
    ...     return "done"
    ...
    >>> t = foo()
    >>> t.isAlive()
    True
    >>> t.join()  # waits for thread completion
    >>> t.isAlive()
    False
    """
    def run(*k, **kw):
        t = threading.Thread(target=fn, args=k, kwargs=kw)
        t.start()
        return t
    return run


class ScriptBase(object):
    """A base class which can be used with import scripts.
    It provides facilities to:
    
     - print progress
     - print results (coloured)
     - gracefully add skipped imports during script iteration
     - write files in a safe manner (in case of errors, files are removed
       before exiting the interpreter)
     - compress generated files
    """
    compress_on_exit = True
    remove_output_dir_on_start = True

    def __init__(self):
        self.total = 0
        self._skipped = 0
        self._imported = 0
        self._exited = False
        self._lowercase_names = set()
        self._skip_reasons = []
        self._started = time.time()
        if self.remove_output_dir_on_start:
            self.safe_remove('out')
        if not os.path.isdir('out'):
            os.mkdir('out')       

    def __del__(self):
        if not self._exited:
            self._tear_down()

    def _tear_down(self):
        self._exited = True
        elapsed = "%0.3f" % (time.time() - self._started)
        hl = hilite
        print "total:%s imported:%s skipped=%s in %s secs" \
           % (hl(self.total), hl(self._imported), hl(self._skipped), hl(elapsed))
        if self._skip_reasons:
            print "skip reasons:"
            for x in set(self._skip_reasons):
                print "(%s) %s" % (hl(self._skip_reasons.count(x), 0), x)
        if self.compress_on_exit:
            self._compress_output_files()

    def skip(self, reason=""):
        """Adds a message to the skip queue shown on exit."""
        self._skipped += 1
        if reason:
            self._skip_reasons.append(reason)
            
    def name_already_processed(self, name):
        """Return True if the name of the person has already been
        processed to avoid duplicate persons.
        """
        if not name:
            return True
        lowercased_name = name.lower()
        if lowercased_name in self._lowercase_names:
            return True
        else:
            self._lowercase_names.add(lowercased_name)
            return False
        
    def print_progress(self, index, name=None):
        s = "processing: %s/%s" %(index, self.total)
        if name is not None:
            s += " - " + repr(name)
        print s
       
    def write_file(self, bdes, id):
        #index = str(index).zfill(len(str(self.total)))
        filename = os.path.join('out', "%s.xml" % id)
        try:
            bdes.to_file(filename)
        except:
            self.safe_remove(filename)
            raise
        self._imported += 1
        
        #from lxml import etree
        #namestring = bdes.get_namen()[0].to_string()
        #etree.fromstring(namestring)

    @classmethod
    def safe_remove(self, path):
        if os.path.isdir(path):
            shutil.rmtree(path, ignore_errors=True)
        else:
            try:
                os.remove(path)
            except OSError:
                pass
    @staticmethod
    def _compress_output_files():
        tar = tarfile.open("out.tar.gz", "w:gz")
        for name in os.listdir("out"):
            tar.add("out/" + name, arcname=name)
        tar.close()


